dpcpp/matrix/bccoo_kernels.dp.cpp

```diff
@@ -0,0 +1,1241 @@
+                                                      start_in_blk, jump_in_blk,
+                                                      block_size_local, diag);
+            } else {
+                loop_block_single_row_extract<uint32>(chunk_data, blk_idxs,
+                                                      start_in_blk, jump_in_blk,
+                                                      block_size_local, diag);
+            }
+        }
+    }
+}
+
+GKO_ENABLE_DEFAULT_HOST(abstract_extract, abstract_extract);
+
+
+/**
+ * The device function of BCCOO compute_absolute_inplace
+ *
+ * @param nnz  the number of nonzeros in the matrix
+ * @param num_blks  the number of blocks in the matrix
+ * @param block_size  the number of nonzeros in each block
+ * @param num_lines  the maximum round of each warp
+ * @param chunk_data  the array where the data are
+ * @param offsets_data  the array where the offset of each block is
+ * @param types_data  the array where the type of each block is
+ * @param cols_data  the array where the initial column of each block is
+ * @param rows_data  the array where the initial row of each block is
+ * @param comp_abs  the function to apply to obtain the solution
+ *
+ * @tparam subgroup_size   size of the subgroup
+ * @tparam ValueType  type of values stored in the matrix
+ * @tparam IndexType  type of matrix indexes stored in the structure
+ * @tparam Closure  type of the function used to write the result
+ */
+template <int subgroup_size = config::warp_size, typename ValueType,
+          typename IndexType, typename Closure>
+void absolute_inplace_kernel(const IndexType nnz, const IndexType num_blks,
+                             const IndexType block_size,
+                             const IndexType num_lines,
+                             uint8* __restrict__ chunk_data,
+                             const size_type* __restrict__ offsets_data,
+                             const uint8* __restrict__ types_data,
+                             const IndexType* __restrict__ cols_data,
+                             const IndexType* __restrict__ rows_data,
+                             Closure comp_abs, sycl::nd_item<3> item_ct1)
+
+{
+    const IndexType column_id = item_ct1.get_group(1);
+    const IndexType start_blk = item_ct1.get_group(2);
+    const IndexType jump_blk = item_ct1.get_group_range(2);
+
+    const IndexType start_in_blk =
+        item_ct1.get_local_id(1) * subgroup_size + item_ct1.get_local_id(2);
+    const IndexType jump_in_blk =
+        item_ct1.get_local_range().get(1) * subgroup_size;
+
+    for (IndexType blk = start_blk; blk < num_blks; blk += jump_blk) {
+        IndexType block_size_local =
+            std::min(block_size, nnz - block_size * blk);
+        compr_idxs<IndexType> idxs(blk, offsets_data[blk]);
+        compr_blk_idxs<IndexType> blk_idxs(
+            rows_data, cols_data, block_size_local, idxs, types_data[blk]);
+        if (blk_idxs.is_multi_row()) {
```

Comment:
the inplace one can use one function like cuda_hip, right?

Reply:
Done!!